cache/cache_rep.h

```diff
@@ -0,0 +1,53 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+//
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+
+#pragma once
+
+#include <atomic>
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <mutex>
+#include <vector>
+
+#include "rocksdb/cache.h"
+#include "rocksdb/slice.h"
+
+// [DO-NOT-COMMIT] [A header inclusion issue][A temporary hack]
+// Including the following two dependencies in cache_rep.h leads to a wall of compiler errors when including cache_rep.h in write_buffer_manager.h
+// It is suspected to be circular dependency issue
+// Therefore I commented these two dependencies out and redefine the const from these two dependencies in this file [DO-NOT-COMMIT] 
+// #include "table/block_based/block_based_table_reader.h"
+// #include "util/coding.h"
+
+namespace ROCKSDB_NAMESPACE {
+class CacheRep {
```

Comment:
[non-blocking]
question 1: what is the role of such a CacheRep class when being used by other dummy entries clients?

my thought so far: my current implementation basically follows the previous design of the CacheRep as an inner private class in WBM. But now since I have more c++ class knowledge, I start to question if this is what we want.

Do we want this CacheRep to be (1)just a C-style struct that gathers different values without any encapsulation (everything being public to the CacheRep user) OR (2)a class that provides encapsulation that could take care (e.g, encapsulate) some dummy entry insertion/release stuff? 

It seems to me that the previous CacheRep design lean towards (1) But now since other dummy entry clients are going to reuse the class, I wonder if (2) is what we want. If (2) is what we want, a much more careful design is needed for the interface and a bit more work in the WBM side than now.

Reply:
Resolved: have some thoughts

Reply:
Sorry for unresolving/resolving; GitHub has a confusing UI - in some places text box for reply is there, in other places it's mysteriously missing. So I clicked around randomly to try to get it back.

> Do we want this CacheRep to be (1)just a C-style struct that gathers different values without any encapsulation (i.e, everything being public to the CacheRep user; CacheRep gets deeply coupled into the user logic) OR (2)a class that provides encapsulation for dummy entry insertion/release/lock/bookkeeping/key generation stuff?

I agree (2) sounds better. My prediction is other clients use it in the same way so logic like `ReserveMemWithCache()` would be duplicated. Although there are differences like you pointed out, where `WriteBufferManager` does that fancy technique to reduce future insertions by delaying current removals, while that isn't needed for our simple dictionary buffer that does all the insertions followed by all the removal followed by nothing. I expect it does not hurt to have that logic present though. If a future client doesn't like it we can add a flag to disable it.

> (a)Rename general terms e.g, CacheRep to specific terms e.g, CacheRepForUsingDummyEntry
-or maybe we need to do this even in this current implementation.

Haven't read the code yet, but what about something like `CacheReservationManager`? I never knew what Rep meant here anyways.

> -InsertDummyEntry(..)/ReleaseDummyEntry(..) need more thinking to make sure something like these can be helpful for both WBM (gradual insertion/release of some fixed-sized dummy entries) and BlockBasedTableBuilder (insertion of variable-sized dummy entries/all-at-once release of dummy entries accumulated so far) and future dummy entry clients (e.g, Bloom/Ribbon filter building buffer).
-maybe an extra ReleaseAllDummyEntry() for BlockBasedTableBuilder.

Sounds good, seems we are aligned. Or add a flag to `FreeMemWithCache()`  (like `Cache::Release()`'s force flag) that forces immediate deletion, and invoke it with the full dictionary buffer size.

> -this needs some thinking into how CacheRep user interacts with CacheRep for dummy entry stuff. If I understand correctly, one WBM instance owns one CacheRep instance till the end of that WBM instance and that CacheRep instance is not owned by other WBM instances at the same time (through unique_ptr in WBM). Same for BlockBasedTableBuilder. Then when a CacheRep instance gets destructed, it is indicated that the CacheRep user (e.g, WBM or BlockBasedTableBuilder) no longer needs to refer to the remaining unreleased dummy entries (e.g, when WBM or BlockBasedTableBuilder themselves get destructed). Therefore the CacheRep can signal to release those remaining dummy entries.

I think it's correct. I wonder whether it's a valid case that there are remaining dummy entries at destruction time or if that cleanup is just defensive programming. Anyways it's fine to keep.